### PR TITLE
POC of turbo-json-parse integration

### PR DESCRIPTION
**Pull request by dnlup**

### CODE REVIEW

File: plugin.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+const compilers = new Map()
+
+function plugin (fastify, options, next) {
+  fastify.addHook('onRoute', (routeOptions) => {
+    const { url } = routeOptions
+    const { body } = routeOptions.schema || {}
+    if (body) {
+      if (!compilers.has(url)) {
+        compilers.set(url, compile(body, options))
+      }
```
Comment (mcollina) [inline plugin.js]: You also need to consider the http method.

File: plugin.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    }
+  })
+
+  fastify.addContentTypeParser('application/json', { parseAs: 'string' },
+    (request, body, done) => {
+      try {
+        const { url, body } = request
+        const parse = compilers.get(url)
```
Comment (mcollina) [inline plugin.js]: if there is no compiler, you should fallback to https://github.com/hapijs/bourne (check the default from fastify)

File: plugin.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+const fp = require('fastify-plugin')
+const compile = require('turbo-json-parse')
+
+const compilers = new Map()
```
Comment (mcollina) [inline plugin.js]: This should be moved inside the plugin function.

**Review by mcollina — COMMENTED** [review]
How are the benchmarks going? Does this improve stuff?

It would need some unit test hardening.

Also, we might refactor certain things in Fastify so you don't have to rely on an external Map.

File: plugin.js (hunk 1 above)
Comment (dnlup) [reply plugin.js]: You are right

File: plugin.js (hunk 2 above)
Comment (dnlup) [reply plugin.js]: Yes I saw bourne in the source code, will add it

File: plugin.js (hunk 3 above)
Comment (dnlup) [reply plugin.js]: Yes, otherwise every plugin invocation will share the same map

**Comment by mcollina** [comment]
Unfortunately this approach is not going to work with wildcards and parametric routes.

What you can do is the following:

* use `onRoute` as you are, but add the compiled function as `opts.config.tjpFunc`.
* use `onRequest` to read that property from `reply.context.conifg.tipFunc` and set it to `req.raw.tipFunc`
* access that property from the content parser itself.

Unfortunately this will cause some of the functions to become megamorphic :/. I think we should support this better in Fastify, but it would be a great start anyway.

**Comment by dnlup** [comment]
@mcollina I will post benchmarks, I found an error.

File: plugin.js (hunk 2 above)
Comment (dnlup) [reply plugin.js]: Is there a way to get it from fastify itself or I have to add it separately?

Comment (mcollina) [reply plugin.js]: You'd have to add it seperately for now.

**Comment by dnlup** [comment]
@mcollina So here are the benchmarks at the 4th run:

**plugin**
```bash
$ npm run benchmark
```

Screenshot from 2019-03-29 16-49-58

**core**
```bash
$ npm run benchmark:core
```

Screenshot from 2019-03-29 16-50-58

There is a performance improvement :tada: .

I'll try to follow your suggestion.

Since you mentioned refectoring in fastify, I think that another thing to consider is how to handle shared schemas and the fact that `turbo-json-parse` does not handle `$refs`. A complete schema has to be supplied to the `compile` function. As of now, if the schema contains shared parts, fastify will populate the common part ***only*** if the shared schema is declared in the `$replace-way`. Also this population will not take place in the `onRoute` hook. Schema resolution and population could be done at plugin level also, but I think I would be duplicating functionality and possibly introducing bugs.

**Comment by mcollina** [comment]
> Since you mentioned refectoring in fastify, I think that another thing to consider is how to handle shared schemas and the fact that turbo-json-parse does not handle $refs. A complete schema has to be supplied to the compile function. As of now, if the schema contains shared parts, fastify will populate the common part only if the shared schema is declared in the $replace-way. Also this population will not take place in the onRoute hook. Schema resolution and population could be done at plugin level also, but I think I would be duplicating functionality and possibly introducing bugs.

That part is really sensitive and I'm not super convinced we can do what you'd need. The best approach would be to resolve things here as well. You might check if it is something that AJV would do for you.

**Comment by dnlup** [comment]
@mcollina Tried to follow your input. Benchmarks (4th run)

```bash
$ npm run benchmark
```

Screenshot from 2019-04-02 11-34-57

```bash
$ npm run benchmark:core
```
Screenshot from 2019-04-02 11-35-45

**Comment by mcollina** [comment]
Good work! How are the benchmarks?

**Comment by mcollina** [comment]
You should add some unit testing, this is very good!

You should also try with a bigger JSON - I expect the gap to increase.

**Comment by dnlup** [comment]
@mcollina I think that to pass a bigger json I should write a dedicated `.js` file where I run `autocannon` programmatically

**Comment by mcollina** [comment]
@dnlup you can load it from a file on disk.

**Comment by dnlup** [comment]
@mcollina After generating a big payload the performance of the plugin is worst than core.
Benchmarks at 4th run:
```bash
$ npm run benchmark
```
plugin

```bash
$ npm run benchmark:core
```
core

The payload is an object with 500 keys of 150 characters strings.

Maybe I have not grasped very well turbo json configuration options.

**Comment by mcollina** [comment]
Try these options:

```js
{
  validate: false, // set to false to disable extra type validation
  validateStrings: false, // set to false to disable extra type validation
  prettyPrint: true
}
```

**Comment by dnlup** [comment]
@mcollina Unfortunately that does not change much (I have not committed this change btw).
Benchmarks at 4th run:

```bash
$ npm run benchmark
```
plugin

```bash
$ npm run benchmark:core
```

core

I also have trouble with the stringified JSON payloads. If the JSON is not pretty printed with options `JSON.stringify(payload, null, 2)` turbo json throws an error saying that it has encountered an unexpected key.

**Comment by dnlup** [comment]
@mcollina Even with unsafe parsing
```js
{
    validate: false,
    validateStrings: false,
    fullMatch: false,
    prettyPrinted: true
  }
```

Benchmarks 4th run

```bash
$ npm run benchmark
```
plugin

```bash
$ npm run benchmark:core
```
core

**Comment by mcollina** [comment]
Would you mind generating a flamegraph with https://clinicjs.org/flame/, and use clinic upload to share it in here

**Comment by dnlup** [comment]
No problem, I think I could also try to parse the payload as buffer.

**Comment by dnlup** [comment]
@mcollina Hope I did everything right, here's the link
https://upload.clinicjs.org/public/5f688232aeeab153e2e002c2037396e0ca9df1a2fbfd745aa1404a1f5c346511/5961.clinic-flame.html#selectedNode=907&zoomedNode=&exclude=83c0&merged=true

Here's the command I have run
```bash
$ clinic flame --on-port 'npm run benchmark:request' -- node benchmarks/tjp.js
```

turbo json parser options

```js
{
    validate: false,
    validateStrings: false,
    prettyPrinted: true
}
```
( I'll commit and push this change now)

**Comment by mcollina** [comment]
Ouch.

@mafitonsh would you like to chime in?

Overall this module is ok, you can land and release. Unfortunately it's faster with small json, but worse on bigger jsons.

**Comment by dnlup** [comment]
@mcollina Ok, I will flag it as experimental given that not all JSON schema features are supported for now. This can be used to try things in the future.

**Pull request closed without merging**